packages/concerto-tools/lib/codegen/fromJsonSchema/cto/jsonschemavisitor.js

```diff
@@ -0,0 +1,729 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const {
+    LocalReference,
+    Reference,
+    ArrayProperty,
+    Property,
+    Properties,
+    NonEnumDefinition,
+    EnumDefinition,
+    Definition,
+    Definitions,
+} = require('./jsonSchemaClasses.js');
+
+/**
+ * Convert the contents of a JSON Schema file to a Concerto JSON model.
+ * Set the following parameters to use:
+ * - metaModelNamespace: the current metamodel namespace.
+ * - namespace: the desired namespace of the generated model.
+ *
+ * @private
+ * @class
+ */
+class JsonSchemaVisitor {
+    /**
+     * Returns true if the property maps to the Concerto DateTime type.
+     * @param {Object} property - a JSON Schema model property.
+     *
+     * @return {Boolean} "true" if the property maps to the Concerto DateTime
+     * type.
+     * @private
+     */
+    isPropertyDateTime(property) {
+        return property.body.format &&
+            (
+                property.body.format === 'date-time' ||
+                property.body.format === 'date'
+            );
+    }
+    /**
+     * Returns true if the property contains an "anyOf" or a "oneOf" element.
+     * @param {Object} property - a JSON Schema model property.
+     *
+     * @return {Boolean} "true" if the property contains aan "anyOf" or a
+     * "oneOf" element.
+     * @private
+     */
+    doesPropertyContainAnyOfOrOneOf(property) {
+        return !!(property.body.anyOf || property.body.oneOf);
+    }
+    /**
+     * Flatten a property containing an "anyOf" or a "oneOf" element.
+     * @param {Object} property - a JSON Schema model property.
+     *
+     * @return {Object} a JSON Schema model property with the "anyOf" or
+     * "oneOf" elements resolved.
+     * @private
+     */
+    flattenAnyOfOrOneOfInProperty(property) {
+        // eslint-disable-next-line no-console
+        console.warn(
+            `Keyword '${
+                property.body.anyOf ? 'anyOf' : 'oneOf'
+            }' in definition '${
+                property.path[property.path.length - 1]
+            }' is not fully supported. Defaulting to first alternative.`
+        );
+
+        return {
+            ...Object.fromEntries(
+                Object.entries(property.body)
```

Comment:
Why do we need a `Object.fromEntries(Object.entries(...))`? Isn't this an identity operation?

The spread operator will clone the object for you?